src/disco/bank/fd_txncache.c

```diff
@@ -0,0 +1,741 @@
+                            hash (starting between 0 and len(hash)-20, a/k/a 44 for signatures, and 12 for hashes).
+
+                            In an unfortunate turn, the offset is also propogated to peers via. snapshot responses,
+                            which only communicate the offset and the respective 20 bytes.  To make sure we are
+                            deduplicating incoming transactions correctly, we must replicate this system even though
+                            it would be easier to just always take the first 20 bytes.  For transactions that we
+                            insert into the cache ourselves, we do just always use a key_offset of zero, so this is
+                            only nonzero when constructed form a peer snapshot. */
+
+  uint  heads[ FD_TXNCACHE_BLOCKCACHE_MAP_CNT ]; /* The hash table for the blockhash.  Each entry is a pointer to the head of a
+                                                    linked list of transactions that reference this blockhash.  As we add
+                                                    transactions to the bucket, the head pointer is updated to the new item, and
+                                                    the new item is pointed to the previous head. */
+
+  ushort pages_cnt;      /* The number of txnpages currently in use to store the transactions in this blockcache. */
+  uint * pages;          /* A list of the txnpages containing the transactions for this blockcache. */
+};
+
+typedef struct fd_txncache_private_blockcache fd_txncache_private_blockcache_t;
+
+struct fd_txncache_private_slotblockcache {
+  uchar blockhash[ 32 ]; /* The actual blockhash of these transactions. */
+  ulong txnhash_offset;  /* As described above. */
+  uint  heads[ FD_TXNCACHE_SLOTCACHE_MAP_CNT ]; /* A map of the head of a linked list of tansactions in this slot and blockhash */
+};
+
+typedef struct fd_txncache_private_slotblockcache fd_txncache_private_slotblockcache_t;
+
+struct fd_txncache_private_slotcache {
+  ulong                                slot; /* The slot that this slotcache is for. */
+  fd_txncache_private_slotblockcache_t blockcache[ 300UL ];
+};
+
+typedef struct fd_txncache_private_slotcache fd_txncache_private_slotcache_t;
+
+struct __attribute__((aligned(FD_TXNCACHE_ALIGN))) fd_txncache_private {
+  fd_rwlock_t lock[ 1 ]; /* The txncache is a concurrent structure and will be accessed by multiple threads
+                            concurrently.  Insertion and querying only take a read lock as they can be done
+                            lockless but all other operations will take a write lock internally. */
+
+  ulong  root_slots_max;
+  ulong  live_slots_max;
+  ushort txnpages_per_blockhash_max;
+
+  ulong   root_slots_cnt; /* The number of root slots being tracked in the below array. */
+  ulong * root_slots; /* The highest N slots that have been rooted.  These slots are
```

Comment:
Could you add a comment about the way this is used? I was definitely expecting something more like a ring buffer with far fewer memmoves, and that initially confused me.